curator-drivers/open-telemetry/src/main/java/org/apache/curator/drivers/opentelemetry/OpenTelemetryTracingDriver.java

```diff
@@ -0,0 +1,185 @@
+ * For example, it will not register attributes if they have the default value, only if the value has diverged from it.
+ */
+public class OpenTelemetryTracingDriver extends AdvancedTracerDriver {
+
+    interface AttributeKeys {
+        AttributeKey<String> PATH_ATTRIBUTE = AttributeKey.stringKey("path");
+        AttributeKey<Long> SESSION_ID_ATTRIBUTE = AttributeKey.longKey("sessionId");
+        AttributeKey<Long> LATENCY_MS_ATTRIBUTE = AttributeKey.longKey("latencyMs");
+        AttributeKey<Long> REQUEST_BYTES_LENGTH_ATTRIBUTE = AttributeKey.longKey("requestBytesLength");
+        AttributeKey<Long> RESPONSE_BYTES_LENGTH_ATTRIBUTE = AttributeKey.longKey("responseBytesLength");
+        AttributeKey<Boolean> WITH_WATCHER_ATTRIBUTE = AttributeKey.booleanKey("withWatcher");
+        AttributeKey<Long> STAT_AVERSION_ATTRIBUTE = AttributeKey.longKey("stat.Aversion");
+        AttributeKey<Long> STAT_CTIME_ATTRIBUTE = AttributeKey.longKey("stat.Ctime");
+        AttributeKey<Long> STAT_CVERSION_ATTRIBUTE = AttributeKey.longKey("stat.Cversion");
+        AttributeKey<Long> STAT_CZXID_ATTRIBUTE = AttributeKey.longKey("stat.Czxid");
+        AttributeKey<Long> STAT_DATA_LENGTH_ATTRIBUTE = AttributeKey.longKey("stat.DataLength");
+        AttributeKey<Long> STAT_EPHEMERAL_OWNER_ATTRIBUTE = AttributeKey.longKey("stat.EphemeralOwner");
+        AttributeKey<Long> STAT_MTIME_ATTRIBUTE = AttributeKey.longKey("stat.Mtime");
+        AttributeKey<Long> STAT_MZXID_ATTRIBUTE = AttributeKey.longKey("stat.Mzxid");
+        AttributeKey<Long> STAT_NUM_CHILDREN_ATTRIBUTE = AttributeKey.longKey("stat.NumChildren");
+        AttributeKey<Long> STAT_PZXID_ATTRIBUTE = AttributeKey.longKey("stat.Pzxid");
+        AttributeKey<Long> STAT_VERSION_ATTRIBUTE = AttributeKey.longKey("stat.Version");
+    }
+
+    private final Tracer tracer;
+
+    public OpenTelemetryTracingDriver(Tracer tracer) {
+        Preconditions.checkNotNull(tracer);
+        this.tracer = tracer;
+    }
+
+    @Override
+    public Object startTrace(OperationTrace trace) {
+        Preconditions.checkNotNull(trace);
+        Span span = tracer.spanBuilder(trace.getName())
+                .setStartTimestamp(trace.getStartTimeNanos(), NANOSECONDS)
+                .startSpan();
+
+        if (trace.getPath() != null) {
+            span.setAttribute(PATH_ATTRIBUTE, trace.getPath());
+        }
+        if (trace.isWithWatcher()) {
+            span.setAttribute(WITH_WATCHER_ATTRIBUTE, true);
+        }
+        if (trace.getSessionId() >= 0) {
+            span.setAttribute(SESSION_ID_ATTRIBUTE, trace.getSessionId());
+        }
+        if (trace.getRequestBytesLength() > 0) {
+            span.setAttribute(REQUEST_BYTES_LENGTH_ATTRIBUTE, trace.getRequestBytesLength());
+        }
+        // We will close this in endTrace
+        @SuppressWarnings("MustBeClosedChecker")
+        Scope scope = span.makeCurrent();
```

Comment:
Curator invokes callbacks asynchronously.

The timing of callback invocation is indeterministic.  `Span::makeCurrent` in `startTrace` and `Scope::close` in `endTrace` probably will not form try-with-resources as suggested in javadoc of `Span::makeCurrent`.

Also, `endTrace` is probably not invoked in thread where `startTrace` called as far as I known. Given `CuratorFramework::getChildren` as an example, `startTrace` is invoked in Curator threads while `endTrace` is invoked in ZooKeeper thread.

Since `startTrace` and `endTrace` are the only interceptions, I think we can drop `scope`/`context` propagation here.